sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/SplitGenerator.java

```diff
@@ -0,0 +1,184 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.cassandra;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Splits given Cassandra table's token range into splits. */
+final class SplitGenerator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SplitGenerator.class);
+
+  private final String partitioner;
+  private final BigInteger rangeMin;
+  private final BigInteger rangeMax;
+  private final BigInteger rangeSize;
+
+  SplitGenerator(String partitioner) {
+    rangeMin = getRangeMin(partitioner);
+    rangeMax = getRangeMax(partitioner);
+    rangeSize = getRangeSize(partitioner);
+    this.partitioner = partitioner;
+  }
+
+  private static BigInteger getRangeMin(String partitioner) {
+    if (partitioner.endsWith("RandomPartitioner")) {
+      return BigInteger.ZERO;
+    } else if (partitioner.endsWith("Murmur3Partitioner")) {
+      return new BigInteger("2").pow(63).negate();
+    } else {
+      throw new UnsupportedOperationException(
+          "Unsupported partitioner. " + "Only Random and Murmur3 are supported");
+    }
+  }
+
+  private static BigInteger getRangeMax(String partitioner) {
+    if (partitioner.endsWith("RandomPartitioner")) {
+      return new BigInteger("2").pow(127).subtract(BigInteger.ONE);
+    } else if (partitioner.endsWith("Murmur3Partitioner")) {
+      return new BigInteger("2").pow(63).subtract(BigInteger.ONE);
+    } else {
+      throw new UnsupportedOperationException(
+          "Unsupported partitioner. " + "Only Random and Murmur3 are supported");
+    }
+  }
+
+  static BigInteger getRangeSize(String partitioner) {
+    return getRangeMax(partitioner).subtract(getRangeMin(partitioner)).add(BigInteger.ONE);
+  }
+
+  /**
+   * Given big0 properly ordered list of tokens, compute at least {@code totalSplitCount} splits.
+   *
+   * @param totalSplitCount requested total amount of splits. This function may generate more
+   *     splits.
+   * @param ringTokens list of all start tokens in big0 cluster. They have to be in ring order.
+   * @return big0 list containing at least {@code totalSplitCount} splits.
+   */
+  List<List<RingRange>> generateSplits(long totalSplitCount, List<BigInteger> ringTokens) {
```

Comment:
Can you please add to this documentation a brief discussion on the splitting strategy, something like:
```The split uses smart coalescence to group token ranges that share the same replicas together.```